### ISSUE TITLE
Add RNS Support

### ISSUE THREAD

**Issue by AlexanderViand-Intel** [google/heir#1140]
In order to support non-trivial FHE programs, HEIR needs to become properly RNS-aware.

See #883 for (part of) the discussion that led to the current (final) design.

We model RNS as simply another `coefficientType`  for polynomials, so instead of 
```llvm
!Z463187969_i32_ = !mod_arith.int<463187969 : i32>
#ring = #polynomial.ring<coefficientType = !Z463187969_i32_, polynomialModulus = <1 + x**8>>
```

We would have (obviously not a real set of RNS primes):
```llvm
!Z463187969_i32_ = !mod_arith.int<463187969 : i32>
!Z463187970_i32_ = !mod_arith.int<463187970 : i32>
!Z463187971_i32_ = !mod_arith.int<463187971 : i32>
!rns = !rns.rns<!Z463187969_i32_, !Z463187970_i32_, !Z463187971_i32_>
#ring = #polynomial.ring<coefficientType = !rns, polynomialModulus = <1 + x**8>>
```
(Note that the above IR is already valid, thanks to #1114)

In order to actually make use of RNS types in the compilation, though, there are several open TODOs:

- [ ]  Replace the `coefficient-mod-bits` flag in the `--secret-to-<RLWE Scheme>` passes with a new (optional) flag that takes the actual RNS primes. Ideally, those should also be verified (or we could outsource this to the `!rns.rns` type verifier?). 
- [ ]  Update the `-to-openfhe-<scheme>` passes? This might not actually be necessary, as it currently just happily forwards the RLWE types with RNS, but it feels like those passes _should_ care about this
- [ ] Update the add-client-interface pass to forward the chosen primes to the crypto context/scheme setup.
- [ ] Update the openFHE emitter accordingly

**Comment by ZenithalHourlyRate** [comment]
While implementing the RNS support for `secret-to-<scheme>`, it is found that the `TypeConverter` from `SecretType` to `RlweCiphertextType` can not handle the case where one `Value` should be converted to level 2 and another `Value` should be converted to level 1, as the `TypeConverter` only have the information from the `SecretType` (and the TypeConverter private member), we currently can only convert all these `SecretType` to one specific `CiphertextType`.

To support different type conversion, I think it might be necessary to add an attribute to `SecretType` and former passes should populate it, then the conversion pass can know which level it should convert to.

**Comment by AlexanderViand-Intel** [comment]
Yes, the lack of a context aware type converter in the Dialect Conversion framework is one of the reasons why the current lowerings always need to emit mul and relin together. Rather than modifying secret type, I think the better solution is to handle this logic ourselves, without relying on the existing type converter infrastructure.

**Comment by ZenithalHourlyRate** [comment]
> I think the better solution is to handle this logic ourselves, without relying on the existing type converter infrastructure.

One tricky thing is that even we can manually lower all the secret::GenericOp, in `SecretToBGV` we have a `addStructuralConversionPatterns` that changes the function signature and all other scf stuff. Are we going to implement all them?

**Comment by AlexanderViand-Intel** [comment]
> Anyway I've done it. By _misusing_ the `convertTypeAttribute` in TypeConverter, manually changing the function signature and totally getting rid of the OpAdaptor (manually change the type of the Value) we can get a SecretType lowered to correspondent CiphertextType

Sounds brutal xD Do you have a link to a (draft) PR?

**Comment by j2kun** [comment]
This is done, and the remaining todos for integrating RNS support are handled by other issues
